### PR TITLE
Various plugin changes

### PR DESCRIPTION
**Pull request by satamas**

### CODE REVIEW

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt [hunk 1]
```diff
@@ -43,46 +46,82 @@ class HubPluginManager private constructor() : PluginManager<HubPlugin> {
...
+  private fun loadIcon(pluginDirectory: File, iconUrl: String?): PluginIcon? {
+    return when {
+      iconUrl == null -> null
+      iconUrl.contains("://") || iconUrl.startsWith("/") -> {
+        LOG.warn("Unsupported widget iconUrl: '$iconUrl'")
+        null
+      }
+      else -> {
+        val iconEntry = File(pluginDirectory, iconUrl)
+        if (!iconEntry.exists()) throw IllegalArgumentException("Icon file doesn't exist")
```
Comment (AlexanderPrendota) [inline intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: use `require { }`

Comment (AlexanderPrendota) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: and why IllegalArgumentException? We use PluginCreationFail over here

Comment (AlexanderPrendota) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: but still keep doing this above
```kotlin
    } catch (e: IllegalArgumentException) {
      return PluginCreationFail(HubIconInvalidUrl())
    }
```

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt [hunk 2]
```diff
@@ -43,46 +46,82 @@ class HubPluginManager private constructor() : PluginManager<HubPlugin> {
...
       LOG.info("Unable to read plugin descriptor $DESCRIPTOR_NAME", e)
       return PluginCreationFail(UnableToReadDescriptor(DESCRIPTOR_NAME, e.localizedMessage))
     }
   }
+
+  private fun loadIcon(pluginDirectory: File, iconUrl: String?): PluginIcon? {
+    return when {
+      iconUrl == null -> null
+      iconUrl.contains("://") || iconUrl.startsWith("/") -> {
+        LOG.warn("Unsupported widget iconUrl: '$iconUrl'")
```
Comment (AlexanderPrendota) [inline intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: If we throw an exception when we can not find the icon let's throw an exception here. and the icon becomes not null value

Comment (satamas) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: It's logic from our application. 
If icon url is an external url than it's not an error, we just will not load it. But if icon url point to an element in archive and this element is missing this is an error

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt (hunk 1 above)
Comment (satamas) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: And what am I supposed to do with PluginCreationFail in method that returns icon?

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt (hunk 2 above)
Comment (AlexanderPrendota) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: okay

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt (hunk 1 above)
Comment (AlexanderPrendota) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: Screen Shot 2020-03-24 at 11 17 04 AM
okay, but you dont need the message `"Icon file doesn't exist`

Comment (AlexanderPrendota) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: just use require

File: .idea/encodings.xml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="Encoding">
+    <file url="file://$PROJECT_DIR$/../plugins/services/uploader/src/main/resources/log4j.properties" charset="UTF-8" />
```
Comment (serejke) [inline .idea/encodings.xml]: what is this? please revert the change

File: intellij-plugin-structure/tests/src/test/kotlin/com/jetbrains/plugin/structure/hub/mock/HubPluginMockTest.kt [hunk 4]
```diff
@@ -1,43 +1,20 @@
...
-  }
-
   @Test
   fun `hub plugin`() {
-    val pluginFile = buildZipFile(temporaryFolder.newFile("plugin.zip")) {
-      file(HubPluginManager.DESCRIPTOR_NAME) {
-        mockHubPluginManifestContent
-      }
-    }
+    val pluginFile = URLUtil.urlToFile(this::class.java.getResource("/hub/hub-mock.zip"))
```
Comment (serejke) [inline intellij-plugin-structure/tests/src/test/kotlin/com/jetbrains/plugin/structure/hub/mock/HubPluginMockTest.kt]: Please revert this. We don't want to pollute git repository with binaries. use `buildZipFile`

And this test doesn't check that `HubPlugin.icons` are correct. Use 
```
buildZip(...) {
   dir("images") {
     file("icon.png", SOME_CONTENT)
   }
}
```

and later check that `HubPlugin.icons.single()` has content of SOME_CONTENT

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt [hunk 5]
```diff
@@ -43,46 +46,82 @@ class HubPluginManager private constructor() : PluginManager<HubPlugin> {
...
+      return PluginCreationFail(PluginDescriptorIsNotFound(DESCRIPTOR_NAME))
     }
-    return PluginCreationFail(PluginDescriptorIsNotFound(DESCRIPTOR_NAME))
+    val manifestContent = manifestFile.readText()
+    val manifest = Json(
+        JsonConfiguration.Stable.copy(isLenient = true, ignoreUnknownKeys = true)
+    ).parse(HubPluginManifest.serializer(), manifestContent)
+    val icon = try {
+      loadIcon(pluginDirectory, manifest.iconUrl)
+    } catch (e: IllegalArgumentException) {
```
Comment (serejke) [inline intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: bad code style (unchecked exceptions mustn't be caught), move this check to `loadIcon` method

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/problems/HubPluginErrors.kt [hunk 6]
```diff
@@ -3,6 +3,16 @@ package com.jetbrains.plugin.structure.hub.problems
 import com.jetbrains.plugin.structure.base.problems.InvalidDescriptorProblem
 import com.jetbrains.plugin.structure.hub.HubPluginManager
 
+class HubIconInvalidUrl : InvalidDescriptorProblem(HubPluginManager.DESCRIPTOR_NAME) {
+
+  override val detailedMessage
+    get() = "iconUrl doesn't represent file"
```
Comment (serejke) [inline intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/problems/HubPluginErrors.kt]: consider rewording to "icon is not found by $iconUrl", but not "represent"

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt (hunk 2 above)
Comment (serejke) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: why warning level then? either "info" or no logging at all

File: intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt (hunk 5 above)
Comment (satamas) [reply intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt]: I can't move it to loadIcon, this catch clause is used to return error response if excpetion is thrown during icon loading
